wptrunner/executors/pytestrunner/fixtures.py

```diff
@@ -45,14 +46,79 @@ def setup(request, session):
         def test_something(setup, session):
             assert session.url == "https://example.org"
 
-    The session is closed when the test module goes out of scope by an
-    implicit call to `session.end`.
+    When the test function goes out of scope, any remaining user prompts
+    and opened windows are closed, and the current browsing context is
+    switched back to the top-level browsing context.
+
+    After returning the session to a near-pristine state, the session
+    is also closed by an implicit call to `session.end`.
     """
 
     def __init__(self, client):
         self.client = client
 
-    @pytest.fixture(scope="module")
+        # store handle of original window
+        # so that we do not close it
+        self.original_window = self.client.window_handle
+        self.client.end()
```

Comment:
You’re making a good point. I think you’re right that [closing the session](http://w3c.github.io/webdriver/webdriver-spec.html#delete-session) is supposed to close existing windows (and for some reason doesn’t in Marionette).

The problem I’m trying to solve here is returning the browser to a pristine state when the tests are done. They may have left a number of windows or user prompts open, deliberately or not, that this fixture needs to take care of closing.

I guess this code will have to change so that the original, root window handle is stored immediately after the session is started. I’m struggling to think of a clean way to do that however…

We can either do it by default in [_wdclient.Session.start_](https://github.com/w3c/wdclient/blob/master/webdriver/client.py#L152), or install some sort of callback/on new session handler which may or may not a more elegant solution.

Thoughts?